### ISSUE TITLE
Configuration file without custom subdomain

### ISSUE THREAD

**Issue by Grafikart** [inconshreveable/ngrok#248]
Hi,

I would like to use yml configuration file but I can't set subdomain to false, and the configuration parser set a subdomain based on tunnel name (only for paid subscriptions)

https://github.com/inconshreveable/ngrok/blob/master/src/ngrok/client/config.go#L124

It would be nice being able to do something like that : 

```
tunnels:
  website:
    addr: 80
    proto: http
    subdomain: false
```

**Comment by inconshreveable** [comment]
I agree, this is an open bug and will be fixed soon.

**Comment by gfoiani** [comment]
+1

**Comment by rlevine** [comment]
+1

**Comment by gregory-yet** [comment]
+1

**Comment by patza** [comment]
+1

**Comment by gregory-yet** [comment]
some news about this issue ? @inconshreveable

**Comment by mgaulton** [comment]
+1

**Comment by inconshreveable** [comment]
Working on getting a fix out for this. This is an issue where the ngrok client 'assumes' that you want to set the subdomain of a tunnel to be equal to the name of the tunnel, so it always creates custom subdomains when using the configuration file. Some people rely on this behavior and breaking changes are not possible. There will be some kind of 'version' identifier in the configuration file that you'll have to set to change this behavior.

**Comment by pyhedgehog** [comment]
Is there a workaround? For example downgrade?

**Comment by yozzh** [comment]
+1

**Comment by pospi** [comment]
This bug basically means you can't use the config file with a free plan; which makes running multiple tunnels a pain. Still, that's what bash scripting is for (;

**Comment by pyhedgehog** [comment]
@pospi Could you explain better, how bash scripting can help? You can't run several instances of `ngrok` and you can't run it with config entries. Doesn't this mean that you can't run multiple tunnels at all (it one of them is http/https)?

**Comment by pospi** [comment]
Oh, really? That's not been my experience. I kinda winged it and guessed, if you run two daemons then the latter will bind its info service to port 4041 instead of 4040. Assuming you have an application that's running two proxies for different ports (a webserver on port 3080 and API service on port 3081, say), this should work:

```
killall ngrok

pushd ~/Downloads
  ./ngrok http 3080 > /dev/null &
  ./ngrok http 3081 > /dev/null &

  echo "Waiting for ngrok to warm up..."
  sleep 10s

  proxy1=`curl -s http://localhost:4040/api/tunnels | node -pe "JSON.parse(require('fs').readFileSync('/dev/stdin').toString()).tunnels[0].public_url"`
  proxy2=`curl -s http://localhost:4041/api/tunnels | node -pe "JSON.parse(require('fs').readFileSync('/dev/stdin').toString()).tunnels[0].public_url"`

  echo "Frontend exposed @ ${proxy1}"
  echo "API exposed @ ${proxy2}"
popd
```

Obviously that requires you unpacked ngrok into your downloads folder and have node installed so you can parse the output, but you could use any other JSON parsing utility (like [jq](https://stedolan.github.io/jq/)?) to do that.

**Comment by pyhedgehog** [comment]
From my experience second instance of ngrok fails with:
```Console
Tunnel session failed: Your account is limited to 1 simultaneous client
```

**Comment by pospi** [comment]
Oh, that'd be because I've authed with their servers then. I suppose they don't want people scripting huge amounts of them anonymously, which is fair enough. Try it again after following instructions here- https://ngrok.com/docs#authtoken

**Comment by pyhedgehog** [comment]
However, I've found workaround:
 1. Start `ngrok http <whatever>`.
 2. [Add](https://ngrok.com/docs#start-tunnel) tcp tunnels by `http_proxy= curl -i -H 'content-type: application/json' -d '{"addr":"22","proto":"tcp","name":"ssh"}' http://127.0.0.1:4040/api/tunnels`

**Comment by pyhedgehog** [comment]
@pospi You are obviously not right. They not allowing **any** tunnels for unauthed clients.

**Comment by rdnetto** [comment]
@inconshreveable Is this still being worked on? (Or alternatively, will the v2 client be open-sourced soon?) A simple workaround like using `subdomain: *` to indicate a randomly generated subdomain should be pretty simple to implement, and would have the advantage of not breaking backward compatibility.

**Comment by igarciaolaizola** [comment]
@inconshreveable any update on this?

**Comment by betlab** [comment]
+1

**Comment by Yaytay** [comment]
+1
Note that this affects API usage too, POSTing this:
{"name":"carol","proto":"http","addr":"localhost:8080"}
or this:
`{"name":"carol","proto":"http","addr":"localhost:8080","subDomain":""}`
generates this:
`{"error_code":103,"status_code":502,"msg":"failed to start tunnel","details":{"err":"Custom subdomains are only available for paid plans. Upgrade at https://dashboard.ngrok.com/billing/plan"}}`

@rdnetto's solution is nice and simple (from our point of view).

**Comment by quisse** [comment]
any update on this?

**Comment by AYapejian** [comment]
Consider me a +1 for all these issues, I don't really care what the subdomain is as long as I can get the url outputted in the ngrok.connect callback.

**Comment by aivus** [comment]
Any news?

It's a very simple issue.

**Comment by pyhedgehog** [comment]
Any news?
Please update us on your plans about this issue.

**Comment by NeutronStar408** [comment]
+1

**Comment by laike9m** [comment]
+1

**Comment by suneetk92** [comment]
+1

**Comment by kenji-siu-bitgamelabs** [comment]
The same parameters work on the command line. It only doesn't work when the parameters are saved in a configuration file. The `ngrok` command assumes that the name of the tunnel to be the name of the subdomain, and it could not be disabled. I tried:

```
    subdomain: ""
```

and

```
    subdomain: *
```

and

```
    subdomain:
```

but it didn't work!

**Comment by mehmetg** [comment]
+1

**Comment by igarciaolaizola** [comment]
> @inconshreveable commented on 25 Jun 2015
I agree, this is an open bug and will be fixed soon.

Any update on this?

**Comment by hadifarnoud** [comment]
is this f* issue going to be fixed?

**Comment by ggarber** [comment]
+100

**Comment by inconshreveable** [comment]
This issue is fixed in 2.1.1. It's available for download via the beta channel: https://dl.equinox.io/ngrok/ngrok/beta

Simply not specifying the `subdomain` or `hostname` option will get you a random tunnel defined via the configuration file or API.

It will not be available in the stable channel for a month or so to ease the migration path while people work through updating their configuration files because it's a backwards incompatible change. More information is here: https://ngrok.com/docs#compat-2.1
